From b16f158a152f914a460a09a41bae3b35642e670a Mon Sep 17 00:00:00 2001
From: Andy Li <andy@onthewings.net>
Date: Thu, 27 Apr 2017 12:50:09 +0800
Subject: [PATCH] make it buildable in cygwin

cherry-picked from 3215311f023e55a7db61699144e4319ba1ba44a5
which is already applied in master
---
 CMakeLists.txt         | 2 +-
 cmake/FindAPACHE.cmake | 8 +++++++-
 libs/common/osdef.h    | 6 +++++-
 libs/common/socket.c   | 2 +-
 libs/std/socket.c      | 4 ++--
 libs/std/sys.c         | 2 +-
 vm/neko.h.in           | 6 +++++-
 7 files changed, 22 insertions(+), 8 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8de1702..3361d73 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -886,7 +886,7 @@ if (STATIC_APACHE)
 else()
 	find_package(APACHE REQUIRED)
 	find_package(APR REQUIRED)
-	set(APACHE_LIBRARIES ${APR_LIBRARIES} ${APRUTIL_LIBRARIES})
+	set(APACHE_LIBRARIES ${APR_LIBRARIES} ${APRUTIL_LIBRARIES} ${HTTPD_LIBRARIES})
 	set(APACHE_INCLUDE_DIRS ${APACHE_INCLUDE_DIR} ${APR_INCLUDE_DIR} ${APRUTIL_INCLUDE_DIR})
 endif()
 
diff --git a/cmake/FindAPACHE.cmake b/cmake/FindAPACHE.cmake
index dd4f80d..a7b660d 100644
--- a/cmake/FindAPACHE.cmake
+++ b/cmake/FindAPACHE.cmake
@@ -13,6 +13,12 @@ find_path(APACHE_INCLUDE_DIR
           PATH_SUFFIXES httpd apache apache2
 )
 
+set(HTTPD_NAMES ${HTTPD_NAMES} httpd)
+find_library(HTTPD_LIBRARIES
+      NAMES ${HTTPD_NAMES}
+      PATHS /usr/lib /usr/local/lib
+)
+
 if(NOT DEFINED APACHE_MODULE_DIR)
    find_program(APXS_BIN NAMES apxs apxs2
              PATH_SUFFIXES httpd apache apache2
@@ -29,4 +35,4 @@ include(FindPackageHandleStandardArgs)
 # handle the QUIETLY and REQUIRED arguments and set APACHE_FOUND to TRUE if 
 # all listed variables are TRUE
 find_package_handle_standard_args(APACHE DEFAULT_MSG APACHE_INCLUDE_DIR )
-mark_as_advanced(APACHE_INCLUDE_DIR)
+mark_as_advanced(APACHE_INCLUDE_DIR HTTPD_LIBRARIES)
diff --git a/libs/common/osdef.h b/libs/common/osdef.h
index 1daebb9..ecf728b 100644
--- a/libs/common/osdef.h
+++ b/libs/common/osdef.h
@@ -38,7 +38,11 @@
 #	define OS_BSD
 #endif
 
-#if defined(NEKO_LINUX) || defined(NEKO_MAC) || defined(NEKO_BSD) || defined(NEKO_GNUKBSD)
+#if defined(__CYGWIN__)
+#	define OS_CYGWIN
+#endif
+
+#if defined(NEKO_LINUX) || defined(NEKO_MAC) || defined(NEKO_BSD) || defined(NEKO_GNUKBSD) || defined (OS_CYGWIN)
 #	define OS_POSIX
 #endif
 
diff --git a/libs/common/socket.c b/libs/common/socket.c
index f183971..7232dd4 100644
--- a/libs/common/socket.c
+++ b/libs/common/socket.c
@@ -120,7 +120,7 @@ PHOST phost_resolve( const char *host ) {
 	PHOST ip = inet_addr(host);
 	if( ip == INADDR_NONE ) {
 		struct hostent *h;
-#	if defined(OS_WINDOWS) || defined(OS_MAC)
+#	if defined(OS_WINDOWS) || defined(OS_MAC) || defined(OS_CYGWIN)
 		h = gethostbyname(host);
 #	else
 		struct hostent hbase;
diff --git a/libs/std/socket.c b/libs/std/socket.c
index 7291e20..f1d7bbb 100644
--- a/libs/std/socket.c
+++ b/libs/std/socket.c
@@ -371,7 +371,7 @@ static value host_resolve( value host ) {
 	ip = inet_addr(val_string(host));
 	if( ip == INADDR_NONE ) {
 		struct hostent *h;
-#	if defined(NEKO_WINDOWS) || defined(NEKO_MAC)
+#	if defined(NEKO_WINDOWS) || defined(NEKO_MAC) || defined(NEKO_CYGWIN)
 		h = gethostbyname(val_string(host));
 #	else
 		struct hostent hbase;
@@ -406,7 +406,7 @@ static value host_reverse( value host ) {
 	unsigned int ip;
 	val_check(host,int32);
 	ip = val_int32(host);
-#	if defined(NEKO_WINDOWS) || defined(NEKO_MAC)
+#	if defined(NEKO_WINDOWS) || defined(NEKO_MAC) || defined(NEKO_CYGWIN)
 	h = gethostbyaddr((char *)&ip,4,AF_INET);
 #	else
 	struct hostent htmp;
diff --git a/libs/std/sys.c b/libs/std/sys.c
index 8003d41..e39bdcf 100644
--- a/libs/std/sys.c
+++ b/libs/std/sys.c
@@ -190,7 +190,7 @@ static value set_cwd( value d ) {
 	</doc>
 **/
 static value sys_string() {
-#if defined(NEKO_WINDOWS)
+#if defined(NEKO_WINDOWS) || defined(NEKO_CYGWIN)
 	return alloc_string("Windows");
 #elif defined(NEKO_GNUKBSD)
 	return alloc_string("GNU/kFreeBSD");
diff --git a/vm/neko.h.in b/vm/neko.h.in
index bb9ec1b..7f20644 100644
--- a/vm/neko.h.in
+++ b/vm/neko.h.in
@@ -58,6 +58,10 @@
 #	define NEKO_MINGW
 #endif
 
+#if defined(__CYGWIN__)
+#	define NEKO_CYGWIN
+#endif
+
 #if defined(__i386__) || defined(_WIN32)
 #	define NEKO_X86
 #endif
@@ -70,7 +74,7 @@
 #	define NEKO_64BITS
 #endif
 
-#if defined(NEKO_LINUX) || defined(NEKO_MAC) || defined(NEKO_BSD) || defined(NEKO_GNUKBSD)
+#if defined(NEKO_LINUX) || defined(NEKO_MAC) || defined(NEKO_BSD) || defined(NEKO_GNUKBSD) || defined(NEKO_CYGWIN)
 #	define NEKO_POSIX
 #endif
 
-- 
2.11.0.windows.1

